File: ios/Vocal/Auth/AuthModels.swift
import Foundation

enum ProviderID: String, Codable, CaseIterable, Identifiable, Sendable {
  case password
  case google = "google.com"
  case apple = "apple.com"

  var id: String { rawValue }
  var isUsableOnIOS: Bool { self == .password || self == .apple }
}

struct IdentitySnapshot: Equatable, Sendable {
  let uid: String
  let email: String?
  let emailVerified: Bool
  let providers: Set<ProviderID>
}

struct InstallationMetadata: Codable, Equatable, Sendable {
  let platform: String
  let appVersion: String
  let buildNumber: Int
  let metadataRevision: Int64
  let osVersion: String
  let deviceModel: String?
}

struct InstallationReport: Codable, Equatable, Sendable {
  let installationId: String
  let platform: String
  let appVersion: String
  let buildNumber: Int
  let metadataRevision: Int64
  let osVersion: String
  let deviceModel: String?

  var metadata: InstallationMetadata {
    InstallationMetadata(
      platform: platform, appVersion: appVersion, buildNumber: buildNumber,
      metadataRevision: metadataRevision, osVersion: osVersion, deviceModel: deviceModel)
  }
}

struct AccountProfile: Codable, Equatable, Sendable {
  let id: String
  let displayName: String
  let email: String?
  let emailVerified: Bool
  let providers: [ProviderID]
}

struct VersionTransition: Codable, Equatable, Sendable {
  let appVersion: String
  let buildNumber: Int
  let metadataRevision: Int64
  let observedAt: Date
}

struct RegisteredDevice: Codable, Equatable, Identifiable, Sendable {
  var id: String { installationId }
  let installationId: String
  let platform: String
  let appVersion: String
  let buildNumber: Int
  let metadataRevision: Int64
  let osVersion: String
  let deviceModel: String?
  let firstSeenAt: Date
  let lastSeenAt: Date
  let versionHistory: [VersionTransition]
  var sessionStatus: String? = nil
}

struct DevicePage: Codable, Equatable, Sendable {
  let items: [RegisteredDevice]
  let nextCursor: String?
}

struct PlatformPolicy: Codable, Equatable, Sendable {
  let minimumBuild: Int?
  let latestBuild: Int?
  let downloadUrl: URL?
}

struct PolicyPlatforms: Codable, Equatable, Sendable {
  let android: PlatformPolicy
  let ios: PlatformPolicy
}

struct AppPolicy: Codable, Equatable, Sendable {
  let requireVerifiedEmail: Bool
  let platforms: PolicyPlatforms
  let revision: Int64
  let updatedAt: Date
}

struct ProcessingAccess: Codable, Equatable, Sendable {
  let allowed: Bool
  let reason: String?
  let downloadUrl: URL?
}

struct SessionResponse: Codable, Equatable, Sendable {
  let user: AccountProfile
  let device: RegisteredDevice
  let policy: AppPolicy
  let access: ProcessingAccess
}

struct ProfileSyncResponse: Codable, Equatable, Sendable {
  let user: AccountProfile
  let policy: AppPolicy
}

struct AccountRecoveryRequest: Codable, Equatable, Sendable {
  let id: String
  let status: String
  let reason: String?
  let requestedAt: Date
  let reviewedAt: Date?
  let reviewReason: String?
  let revision: Int64
}

struct AccountRecoveryDeletion: Codable, Equatable, Sendable {
  let requestId: String
  let requestedAt: Date?
  let recoverUntil: Date?
  let phase: String?
  let failureCode: String?
  let recoveryAvailable: Bool
}

struct AccountRecoveryStatus: Codable, Equatable, Sendable {
  let accountStatus: String
  let deletion: AccountRecoveryDeletion?
  let request: AccountRecoveryRequest?
}

enum MailOutcome: String, Codable, Sendable {
  case accepted
  case alreadyVerified = "already_verified"
}

struct MailResponse: Codable, Sendable {
  let status: MailOutcome
}

enum AuthFailure: Error, Equatable, Sendable {
  case cancelled
  case invalidInput
  case invalidCredentials
  case weakPassword
  case accountDisabled
  case accountDeletionPending
  case accountRecoveryExpired
  case collision
  case providerAlreadyLinked
  case providerMissing
  case recentLoginRequired
  case offline
  case rateLimited(retryAt: Date)
  case profileSyncRequired
  case deviceConflict
  case serviceUnavailable
  case sessionExpired
  case configuration
  case malformedResponse
  case unknown

  var messageKey: LocalizedStringKeyName {
    switch self {
    case .cancelled: return "auth_cancelled"
    case .invalidInput: return "auth_invalid_input"
    case .invalidCredentials: return "auth_invalid_credentials"
    case .weakPassword: return "auth_weak_password"
    case .accountDisabled: return "auth_account_disabled"
    case .accountDeletionPending: return "auth_account_deletion_pending"
    case .accountRecoveryExpired: return "auth_account_recovery_expired"
    case .collision: return "auth_collision"
    case .providerAlreadyLinked: return "auth_provider_linked"
    case .providerMissing: return "auth_provider_missing"
    case .recentLoginRequired: return "auth_recent_login"
    case .offline: return "auth_offline"
    case .rateLimited: return "auth_rate_limited"
    case .profileSyncRequired: return "auth_profile_sync_required"
    case .deviceConflict: return "auth_device_conflict"
    case .serviceUnavailable: return "auth_service_unavailable"
    case .sessionExpired: return "auth_session_expired"
    case .configuration: return "auth_configuration"
    case .malformedResponse: return "auth_service_unavailable"
    case .unknown: return "auth_unknown_error"
    }
  }
}

typealias LocalizedStringKeyName = String

extension JSONDecoder {
  static func authDecoder() -> JSONDecoder {
    let decoder = JSONDecoder()
    let fractional = ISO8601DateFormatter()
    fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    let wholeSeconds = ISO8601DateFormatter()
    wholeSeconds.formatOptions = [.withInternetDateTime]
    decoder.dateDecodingStrategy = .custom { decoder in
      let container = try decoder.singleValueContainer()
      let value = try container.decode(String.self)
      guard let date = fractional.date(from: value) ?? wholeSeconds.date(from: value) else {
        throw DecodingError.dataCorruptedError(
          in: container, debugDescription: "Invalid ISO-8601 date")
      }
      return date
    }
    return decoder
  }
}

File: android/app/src/main/java/com/hatem/musicmute/auth/AuthModels.kt
package com.hatem.musicmute.auth

import kotlinx.serialization.Serializable

const val PASSWORD_PROVIDER = "password"
const val GOOGLE_PROVIDER = "google.com"
const val APPLE_PROVIDER = "apple.com"
const val MAX_METADATA_REVISION = 9_007_199_254_740_991L

data class IdentitySnapshot(
    val uid: String,
    val email: String?,
    val emailVerified: Boolean,
    val providers: Set<String>,
    val displayName: String? = null,
)

@Serializable
data class AccountProfile(
    val id: String,
    val displayName: String,
    val email: String? = null,
    val emailVerified: Boolean,
    val providers: List<String>,
)

@Serializable
data class InstallationMetadata(
    val platform: String = "android",
    val appVersion: String,
    val buildNumber: Int,
    val osVersion: String,
    val deviceModel: String? = null,
)

@Serializable
data class InstallationReport(
    val installationId: String,
    val platform: String = "android",
    val appVersion: String,
    val buildNumber: Int,
    val metadataRevision: Long,
    val osVersion: String,
    val deviceModel: String? = null,
) {
    fun metadata() = InstallationMetadata(platform, appVersion, buildNumber, osVersion, deviceModel)
}

@Serializable
data class VersionObservation(
    val appVersion: String,
    val buildNumber: Int,
    val metadataRevision: Long,
    val observedAt: String,
)

@Serializable
data class RegisteredDevice(
    val installationId: String,
    val platform: String,
    val appVersion: String,
    val buildNumber: Int,
    val metadataRevision: Long,
    val osVersion: String,
    val deviceModel: String? = null,
    val firstSeenAt: String,
    val lastSeenAt: String,
    val versionHistory: List<VersionObservation> = emptyList(),
    val sessionStatus: String = "unknown",
)

@Serializable
data class DevicePage(val items: List<RegisteredDevice>, val nextCursor: String? = null)

@Serializable
data class PlatformPolicy(
    val minimumBuild: Int? = null,
)

@Serializable data class PolicyPlatforms(val android: PlatformPolicy, val ios: PlatformPolicy)

@Serializable
data class AppPolicy(
    val requireVerifiedEmail: Boolean,
    val platforms: PolicyPlatforms,
    val revision: Long,
    val updatedAt: String,
)

@Serializable
data class ProcessingAccess(
    val allowed: Boolean,
    val reason: String? = null,
)

@Serializable
data class SessionResponse(
    val user: AccountProfile,
    val device: RegisteredDevice,
    val policy: AppPolicy,
    val access: ProcessingAccess,
)

@Serializable data class ProfileSyncResponse(val user: AccountProfile, val policy: AppPolicy)

@Serializable data class MailOutcome(val status: String)

enum class AuthProblem {
    INVALID_INPUT,
    INVALID_CREDENTIALS,
    EMAIL_IN_USE,
    CREDENTIAL_IN_USE,
    WEAK_PASSWORD,
    REAUTH_REQUIRED,
    ACCOUNT_MISMATCH,
    UNAUTHENTICATED,
    ACCOUNT_DISABLED,
    ACCOUNT_DELETION_PENDING,
    ACCOUNT_RECOVERY_EXPIRED,
    RATE_LIMITED,
    OFFLINE,
    SERVICE_UNAVAILABLE,
    CONFIGURATION,
    STORAGE,
    LAST_METHOD,
    EMAIL_UNAVAILABLE,
    CANCELLED,
    PROFILE_SYNC_REQUIRED,
    DEVICE_CONFLICT,
    GLOBAL_LOGOUT_UNCONFIRMED,
}

class AuthFailure(val problem: AuthProblem, val retryAfterSeconds: Long? = null, val httpStatus: Int? = null) :
    Exception(problem.name)

fun canUnlink(providers: Set<String>, provider: String): Boolean =
    provider in providers &&
        providers.minus(provider).any { it == PASSWORD_PROVIDER || it == GOOGLE_PROVIDER }

fun validAuthEmail(email: String): Boolean =
    email.length in 3..254 &&
        !email.any { it.isWhitespace() || it.isISOControl() } &&
        email.substringBefore('@').isNotEmpty() &&
        email.count { it == '@' } == 1 &&
        email.substringAfter('@').contains('.') &&
        !email.endsWith('.')

@Serializable
data class AccountDeletionReceipt(
    val requestId: String,
    val status: String,
    val recoverUntil: String? = null,
)

@Serializable
data class AccountRecoveryRequest(
    val id: String,
    val status: String,
    val reason: String? = null,
    val requestedAt: String,
    val reviewedAt: String? = null,
    val reviewReason: String? = null,
    val revision: Long,
)

@Serializable
data class AccountRecoveryDeletion(
    val requestId: String,
    val requestedAt: String? = null,
    val recoverUntil: String? = null,
    val phase: String? = null,
    val failureCode: String? = null,
    val recoveryAvailable: Boolean,
)

@Serializable
data class AccountRecoveryStatus(
    val accountStatus: String,
    val deletion: AccountRecoveryDeletion? = null,
    val request: AccountRecoveryRequest? = null,
)
